linkedhashset

yashwanth9090 opened the pull request.

yashwanth9090 committed:
Create linkedhashset

yashwanth9090 committed:
Delete linkedhashset

yashwanth9090 committed:
Create sample

yashwanth9090 committed:
Add files via upload

yashwanth9090 committed:
Create sample

yashwanth9090 committed:
Add files via upload

yashwanth9090 committed:
Delete sample

yashwanth9090 committed:
Delete ArrayIndexCalculator.java

yashwanth9090 committed:
Delete Container.java

yashwanth9090 committed:
Delete HashSet.java

yashwanth9090 committed:
Delete LinkedHashSet.java

yashwanth9090 committed:
Delete MyIterator.java

yashwanth9090 committed:
Delete Node.java

yashwanth9090 committed:
Delete OrderedSet.java

yashwanth9090 committed:
Delete LinkedHashSetTest.java

yashwanth9090 committed:
First Commit

yashwanth9090 committed:
Delete sample

umanchikatla commented on src/com/yashu/interview/map/linkedhashset/HashSet.java:
```diff
@@ -0,0 +1,218 @@
+	// Returns the size of the Array.
+	public int getLengthOfArray() {
+		return array.length;
+	}
```
Remove getLengthOfArray() method. It doesn't need to be public and you get the length by simply calling array.length wherever you need.

umanchikatla commented on src/com/yashu/interview/map/linkedhashset/HashSet.java:
```diff
@@ -0,0 +1,218 @@
+	}
+	
+	//Iterator
+	private class LinkedHashSetIterator implements MyIterator<T>{
```
Implement java iterator instead of MyIterator.

umanchikatla commented on src/com/yashu/interview/map/linkedhashset/HashSet.java:
```diff
@@ -0,0 +1,218 @@
+			if(this.hasNext()){
+				return temp.getData();
+			}
+			return temp.getData();
```
What is the purpose of  if() condition on line 58. It is redundant. You can remove it. 

What happens if I call next() even if hasNext() returns null? It will throw a null pointer exception. 
You should through an exception (NoSuchElementException) as defined in Iterator interface.

umanchikatla commented on src/com/yashu/interview/map/linkedhashset/HashSet.java:
```diff
@@ -0,0 +1,218 @@
+		int indexOfString = hashFunction(value);
+		//If the index is not null, collision is avoided using LinkedList. 
+		if(collision(value)){
+			linkedList.setTail(avoidCollision(value,newNode,linkedList.getTail(),indexOfString));
```
You are already setting LinkedList tail in avoidCollision() and returning the tail. You don't need to set it again.

umanchikatla commented on src/com/yashu/interview/map/linkedhashset/HashSet.java:
```diff
@@ -0,0 +1,218 @@
+		//If the index is not null, collision is avoided using LinkedList. 
+		if(collision(value)){
+			linkedList.setTail(avoidCollision(value,newNode,linkedList.getTail(),indexOfString));
+			count--;
```
If you are only tracking the count of elements inserted in the array, excluding the collided entries, and double array only when it gets full, HashSet performance will be inefficient.  It is very likely that, it will never get full but, lots entries keep getting added to the collision linked list. 

To double the array take into account, total number of elements inserted.

umanchikatla commented on src/com/yashu/interview/map/linkedhashset/HashSet.java:
```diff
@@ -0,0 +1,218 @@
+		for(int i = 0;i<length;i++){
+			temp[i] = array[i];
+		}
+		createNewArray(2*length);
```
You don't need to create a new temp array with existing size. You can simply do...

Node<T>[] temp = array;
createNewArray(2*length);

umanchikatla commented on src/com/yashu/interview/map/linkedhashset/HashSet.java:
```diff
@@ -0,0 +1,218 @@
+		for(int i =0;i<length;i++){
+			hashSetArray[i] = temp[i];
+		}
+		return hashSetArray;
```
- You don't need to create another array with double size after you call createNewArray()
- You can simply use "array" which is empty and double the size 
- Existing values should not be copies into the new array as is. They need to rehashed and inserted. Otherwise, if the same value is inserted again, its hashed index might be different and you will have duplicate values stored

umanchikatla commented on src/com/yashu/interview/map/linkedhashset/HashSet.java:
```diff
@@ -0,0 +1,218 @@
+			linkedList.delete(element);
+			return element.getNext();
+		}else{
+			Node<T> previous = element;
```
previous should be null before the iteration begins

umanchikatla commented on src/com/yashu/interview/map/linkedhashset/HashSet.java:
```diff
@@ -0,0 +1,218 @@
+					previous.setRight(current.getRight());
+				}
+				previous = current;
+				current = current.getNext();
```
Shouldn't you stop looping after you found the element? 

This whole linked list manipulation is more confusing now. Some code is here and some is in LinkedList class.
Lets discuss on phone about how to improve this.

yashwanth9090 committed:
Update ArrayIndexCalculator.java

Throws exception when null object or 0/-ve array length are passed.

yashwanth9090 committed:
Delete Container.java

yashwanth9090 committed:
Delete MyIterator.java

yashwanth9090 committed:
Delete LinkedHashSet.java

yashwanth9090 committed:
Update and rename HashSet.java to LinkedHashSet.java

1. Used java iterator instead of an own iterator and throws an exception when the list is empty.
2. Array size doubles when the linked list size is twice the size of the array.
3. Updated the return statement of delete() to boolean value.

yashwanth9090 committed:
Update LinkedList.java

yashwanth9090 committed:
Update Node.java

yashwanth9090 committed:
Update OrderedSet.java

yashwanth9090 committed:
Update LinkedHashSetTest.java
